File: get-dividends.pl
use DBI;
use Getopt::Long::Descriptive;
use Switch;
use strict;
use warnings;
use List::MoreUtils;
use xfrs;
use Finance::Math::IRR;
use Time::Piece;
use experimental 'smartmatch'; # for using ``if ($var ~~ @array)``

my ($opt, $usage) = describe_options(
  '%c %o',
  [ 'db|d=s',    "Sqlite3 DB file to import into",   { default  => "xfrs.sqlite3.db" } ],
  [ 'order|o=s', "Records order (`date`, `symbol`)", { default  => "date" } ],
  [],
  [ 'help|h',       "print usage message and exit", { shortcircuit => 1 } ],
);


my $dbh = DBI->connect("dbi:SQLite:dbname=".$opt->db,"","",{ RaiseError => 1 }) or die $DBI::errstr;;
my $order = 'date';
if ($opt->order ~~ ['date', 'symbol']) {
    $order = $opt->order;
} else {
    print STDERR "Warning: Unknown ordering ('$opt->order'). Ignored ...\n";
}

my @divs = xfrs::getDividends($dbh,('order' => $order));

foreach my $d (@divs) {
    print "$d->{'symbol'}\t$d->{'currency'}\t$d->{'amount'}\t$d->{'tax'}\t$d->{'date'}\n";
}

$dbh->disconnect();

File: get-dividends.py
import sqlite3
import sys
import xfrs
import argparse


parser = argparse.ArgumentParser(
        description="Prints records of received dividends."
        )

# `db` option: Path to the XFRS database.
parser.add_argument('-d', '--db', type=str,
        default='xfrs.sqlite3.db', dest='db',
        help="Path to the DB file.",
        )

# `order` option: Letting order records by date or by symbol.
parser.add_argument('-o', '--order', type=str, choices = ['date', 'symbol'],
        default='date', dest='order',
        help="By what to order the dividend records",
        )

# `from` option: Letting users select only records beyond this date.
parser.add_argument('-f', '--from', type=str,
        default=None, dest='fromDate',
        help="Return only records on and after this date. Format YYYY-MM-DD.",
        )

# `to` option: Letting users select only records beyond this date.
parser.add_argument('-t', '--to', type=str,
        default=None, dest='toDate',
        help="Return only records by this date. Format YYYY-MM-DD.",
        )

# `symbols` option: Space separated list of stock symbols.
parser.add_argument('-s', '--symbols', type=str,
        default=None, dest='symbols',
        help="Space separated list of symbols to which limit the dividends.",
        )

args = parser.parse_args()

dbh = sqlite3.connect(args.db)

# resolve symbols
symbols = None; # default, means all from DB
if args.symbols is not None:
    symbols = args.symbols.split()

# query dividends
divs = xfrs.getDividends(dbh, symbols, order=args.order, fromDate=args.fromDate, toDate=args.toDate)

dbh.commit()
dbh.close()

for d in divs:
    print(f'{d["symbol"]}\t{d["currency"]}\t{d["amount"]}\t{d["tax"]}\t{d["date"]}')

